In [1]:
import requests
import pandas as pd
import numpy as np
import plotly.express as px
import ipywidgets as widgets
import subprocess
is_colab = subprocess.os.getcwd()=='/content'

In [15]:
from googledrivedownloader import download_file_from_google_drive as gdd
dir(googledrivedownloader.download.download_file_from_google_drive)

['__annotations__',
 '__builtins__',
 '__call__',
 '__class__',
 '__closure__',
 '__code__',
 '__defaults__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__get__',
 '__getattribute__',
 '__getstate__',
 '__globals__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__kwdefaults__',
 '__le__',
 '__lt__',
 '__module__',
 '__name__',
 '__ne__',
 '__new__',
 '__qualname__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__type_params__']

In [18]:
if is_colab:
  !pip install googledrivedownloader
  from googledrivedownloader import download as gdd
  def download_file_g_drive_url(url, file_path, unzip):
    file_id = url[url.index("/d/") + 3: -1*len("/view?usp=sharing")]
    gdd.download_file_from_google_drive(file_id, file_path, unzip)
  energies_path = "/content/energies.csv"
  url = "https://drive.google.com/file/d/1p0SxTqqk01IiKJrqjRavhU6yBhvx7VZB/view?usp=sharing"
  download_file_g_drive_url(url, energies_path, False)
  parent_dir = subprocess.os.getcwd()
else:
  parent_dir = subprocess.os.getcwd()[:-1*len("JupyterNotebooks")]
  energies_path = parent_dir + "ExportedData\\total_energies_of_decay_chains.csv"
energies_path

'/content/energies.csv'

In [19]:
df = pd.read_csv(energies_path)
df

100Ag    100Cd    100In     100Kr  100Nb     100Rb  100Rh    100Sn  \
0  1055.3  1518.64  2857.78  17239.11   2414  13269.11  40.92  2857.78   

    100Sr       100Tc  ...    99Nb   99Pd      99Rb   99Rh     99Sr  99Tc  \
0  9001.1  1317.28575  ...  1863.4  455.8  12758.52  12.63  9104.53  84.6   

       99Y     99Zr  9C   9Li  
0  5910.06  3558.24   0  5710  

[1 rows x 2457 columns]

In [20]:
px.histogram(df.iloc[0,:])

Now to import a list of availablely occuring isotopes.

In [21]:
if is_colab:
    livechart_url = "https://drive.google.com/file/d/1KAarySCi2HCjQCdDrKQoZIaxEOcyTem8/view?usp=sharing"
    livechart_path = "/content/livechart.csv"
    download_file_g_drive_url(livechart_url, livechart_path, False)
else:
    livechart_path = parent_dir + "ImportedData\\all_livechart_decays_and_emissions.csv"
livechart = pd.read_csv(livechart_path)
livechart.loc[:,"A"] = livechart.loc[:,"z"] + livechart.loc[:,"n"]
livechart.loc[:,"Isotope"] = [str(livechart.loc[row,"A"]) + livechart.loc[row,"symbol"]
                                for row in range(livechart.shape[0])]
available_isotopes = livechart.loc[:,"Isotope"].unique()

In [29]:
livechart

Unnamed: 0    z    n symbol   radius   unc_r abundance abundance_unc  \
0               0    0    1     Nn  -0.1149  0.0027                     NaN   
1               1    0    4      N              NaN                     NaN   
2               2    0    6      N              NaN                     NaN   
3               3    1    0      H   0.8783  0.0086   99.9855            78   
4               4    1    1      H   2.1421  0.0088    0.0145            78   
...           ...  ...  ...    ...      ...     ...       ...           ...   
20197        3362  116  177     Lv              NaN                     NaN   
20198        3363  116  178     Lv              NaN                     NaN   
20199        3364  117  176     Ts              NaN                     NaN   
20200        3365  117  177     Ts              NaN                     NaN   
20201        3366  118  176     Og              NaN                     NaN   

      energy_shift energy  ...     atomic_mass unc_am     massexcess unc_me  \
0                       0  ...    1008664.9159     47    8071.318060     44   
1                       0  ...                                   NaN          
2                       0  ...                                   NaN          
3                       0  ...  1007825.031898     14    7288.971064     13   
4                       0  ...  2014101.777844     15   13135.722895     15   
...            ...    ...  ...             ...    ...            ...    ...   
20197                   0  ...       293204583    553  190568.000000    515   
20198                   0  ...                                   NaN          
20199            X         ...       293208727    835  194428.000000    778   
20200                   0  ...       294210840    637  196397.000000    593   
20201                   0  ...       294213979    594  199320.000000    553   

      ENSDFpublicationcut-off                     ENSDFauthors  \
0                 31-Oct-2005                     BALRAJ SINGH   
1                 30-Aug-2017      J.E. Purcell and  C.G. Sheu   
2               15-March-2017       J.H. Kelley and  G.C. Sheu   
3                 31-Oct-2005                     BALRAJ SINGH   
4                  1-May-2003     J.H. KELLEY and  J.L. GODWIN   
...                       ...                              ...   
20197             12-Aug-2005  M. GUPTA and  THOMAS W. BURROWS   
20198             31-Jan-2019                     BALRAJ SINGH   
20199           1-August-2010                         M. GUPTA   
20200             31-Jan-2019                     BALRAJ SINGH   
20201             31-Jan-2019                     BALRAJ SINGH   

      Extraction_date Decay or Emission type    A Isotope  
0          2022-04-26       beta minus decay    1     1Nn  
1          2022-04-26       beta minus decay    4      4N  
2          2022-04-26       beta minus decay    6      6N  
3          2022-04-26       beta minus decay    1      1H  
4          2022-04-26       beta minus decay    2      2H  
...               ...                    ...  ...     ...  
20197      2022-04-26         x ray emission  293   293Lv  
20198      2022-04-26         x ray emission  294   294Lv  
20199      2022-04-26         x ray emission  293   293Ts  
20200      2022-04-26         x ray emission  294   294Ts  
20201      2022-04-26         x ray emission  294   294Og  

[20202 rows x 57 columns]

In [28]:
livechart.iloc[0,:]

Unnamed: 0                                0
z                                         0
n                                         1
symbol                                   Nn
radius                              -0.1149
unc_r                                0.0027
abundance                                  
abundance_unc                           NaN
energy_shift                               
energy                                    0
unc_e                                      
ripl_shift                                 
jp                                     1/2+
half_life                             613.9
operator_hl                                
unc_hl                                    6
unit_hl                                   s
half_life_sec                         613.9
unc_hls                                 0.6
decay_1                                  B-
decay_1_%                               100
unc_1                                      
decay_2                                    
decay_2_%                                  
unc_2                                      
decay_3                                    
decay_3_%                                  
unc_3                                      
isospin                                    
magnetic_dipole                  -1.9130427
unc_md                                     
electric_quadrupole                        
unc_eq                                     
qbm                                 782.347
unc_qb                                    4
qbm_n                                      
unc_qbmn                                NaN
qa                                         
unc_qa                                  NaN
qec                                        
unc_qec                                 NaN
sn                                      0.0
unc_sn                                    0
sp                                         
unc_sp                                  NaN
binding                                   0
unc_ba                                    0
atomic_mass                    1008664.9159
unc_am                                   47
massexcess                       8071.31806
unc_me                                   44
ENSDFpublicationcut-off         31-Oct-2005
ENSDFauthors                   BALRAJ SINGH
Extraction_date                  2022-04-26
Decay or Emission type     beta minus decay
A                                         1
Isotope                                 1Nn
Name: 0, dtype: object

For every available isotope, plot additional energy released by the decay chain after a neutron is added to the nucleus.

In [22]:
def energy(isotope):
    if isotope in df.columns:
        return df.loc[0,isotope]
    else:
        return 0
def add_neutron(isotope):
    element = ''.join([c for c in isotope if not c.isdigit()])
    a = int(''.join([c for c in isotope if c.isdigit()]))
    return str(a+1) + element

In [23]:
post_absorption_energy = {}
for isotope in available_isotopes:
    post_absorption_energy[isotope] = energy(add_neutron(isotope)) - energy(isotope)
post_absorption_energy

{'1Nn': 0,
 '4N': 0,
 '6N': 0,
 '1H': 0,
 '2H': np.float64(5.69),
 '3H': np.float64(2880.39),
 '4H': np.float64(-2886.08),
 '5H': 0,
 '6H': 0,
 '7H': 0,
 '3He': 0,
 '4He': 0,
 '5He': np.float64(1567.6),
 '6He': np.float64(-1567.6),
 '7He': np.float64(10693.95),
 '8He': np.float64(-10693.95),
 '9He': 0,
 '10He': 0,
 '4Li': 0,
 '5Li': 0,
 '6Li': 0,
 '7Li': np.float64(6293.95),
 '8Li': np.float64(-583.9499999999998),
 '9Li': np.int64(-5710),
 '10Li': np.float64(11391.36),
 '11Li': np.float64(-11391.36),
 '12Li': 0,
 '6Be': 0,
 '7Be': np.float64(45.95),
 '8Be': np.float64(-45.95),
 '9Be': np.float64(202.56),
 '10Be': np.float64(4449.889999999999),
 '11Be': np.float64(7323.509999999999),
 '12Be': np.float64(-11975.96),
 '13Be': np.float64(13257.78),
 '14Be': np.float64(-13257.78),
 '15Be': 0,
 '16Be': 0,
 '7B': np.float64(6381.46),
 '8B': np.float64(-6381.46),
 '9B': 0,
 '10B': 0,
 '11B': np.float64(6388.94),
 '12B': np.float64(-36.07999999999993),
 '13B': np.float64(256.6300000000001),
 '1

In [24]:
efficiency_df = pd.DataFrame(data = post_absorption_energy,
            index = ["Change in Energy Release with Neutron Absorption (keV)"])
efficiency_df = efficiency_df.transpose()
efficiency_df

Change in Energy Release with Neutron Absorption (keV)
1Nn                                                 0.00     
4N                                                  0.00     
6N                                                  0.00     
1H                                                  0.00     
2H                                                  5.69     
...                                                  ...     
293Lv                                               0.00     
294Lv                                               0.00     
293Ts                                               0.00     
294Ts                                               0.00     
294Og                                               0.00     

[3367 rows x 1 columns]

Does this account for excited states in the nucleus?

In [25]:
efficiency_df["Energy Release without Neutron Absorption (keV)"] = [
    energy(isotope) for isotope in efficiency_df.index
]
efficiency_df["Parent Isotope"] = [isotope for isotope in efficiency_df.index]
efficiency_df

Change in Energy Release with Neutron Absorption (keV)  \
1Nn                                                 0.00        
4N                                                  0.00        
6N                                                  0.00        
1H                                                  0.00        
2H                                                  5.69        
...                                                  ...        
293Lv                                               0.00        
294Lv                                               0.00        
293Ts                                               0.00        
294Ts                                               0.00        
294Og                                               0.00        

       Energy Release without Neutron Absorption (keV) Parent Isotope  
1Nn                                                0.0            1Nn  
4N                                                 0.0             4N  
6N                                                 0.0             6N  
1H                                                 0.0             1H  
2H                                                 0.0             2H  
...                                                ...            ...  
293Lv                                              0.0          293Lv  
294Lv                                              0.0          294Lv  
293Ts                                              0.0          293Ts  
294Ts                                              0.0          294Ts  
294Og                                              0.0          294Og  

[3367 rows x 3 columns]

In [26]:
fig = px.scatter(efficiency_df, x="Energy Release without Neutron Absorption (keV)",
  y = "Change in Energy Release with Neutron Absorption (keV)",
  hover_data = ["Parent Isotope"],
  title = "Efficiency: Additional Energy Released by Decay Chain after Absorbing 1 Neutron")
fig

Now let's make the x-axis the half-life of the first decay in each chain, which will put a strict upperbound on the specific power of each isotope's decay chain.

In [41]:
livechart.loc[livechart["Isotope"] == '84Nb', "half_life_sec"].unique()

array(['9.8'], dtype=object)

In [63]:
half_lives_dict = {}
isotopes_with_multiple_half_lives = []
isotopes_with_no_half_lives = []
for isotope in available_isotopes:
    half_lives = livechart.loc[livechart["Isotope"] == isotope, "half_life_sec"]
    unique_half_lives = half_lives.unique()
    if len(unique_half_lives) > 1:
        isotopes_with_multiple_half_lives.append(isotope)
    #if there are multiple half lives for an isotope
    #need to add more code to find an expectation value if this is the case
    if not all(unique_half_lives == ' '):
        half_lives_dict[isotope] = np.float64(unique_half_lives[0])
    else:
        isotopes_with_no_half_lives.append(isotope)
half_lives_dict

{'1Nn': np.float64(613.9),
 '4N': np.float64(1.7547604425822527e-22),
 '3H': np.float64(388781328.00697345),
 '5H': np.float64(8.608258774931805e-23),
 '6H': np.float64(2.943469129492811e-22),
 '7H': np.float64(5.069307945237619e-21),
 '5He': np.float64(7.040705479496693e-22),
 '6He': np.float64(0.8067),
 '7He': np.float64(3.0415847671425716e-21),
 '8He': np.float64(0.1191),
 '10He': np.float64(1.5207923835712858e-21),
 '5Li': np.float64(3.709249716027526e-22),
 '8Li': np.float64(0.8399),
 '9Li': np.float64(0.1783),
 '11Li': np.float64(0.00875),
 '6Be': np.float64(4.959105598602019e-21),
 '7Be': np.float64(4598208.0),
 '8Be': np.float64(8.190982317260066e-17),
 '10Be': np.float64(47650958221633.92),
 '11Be': np.float64(13.76),
 '12Be': np.float64(0.02146),
 '13Be': np.float64(2.7e-21),
 '14Be': np.float64(0.00435),
 '15Be': np.float64(7.866167501230788e-22),
 '16Be': np.float64(5.702971438392321e-22),
 '7B': np.float64(5.695851623862494e-22),
 '8B': np.float64(0.77),
 '9B': np.float64(

In [62]:
[len(isotopes_with_no_half_lives),
 len(isotopes_with_multiple_half_lives), len(available_isotopes)]

[415, 0, 3367]

So although we're missing 415 out of the 3367 half-lives, none of the isotopes in this version of livechart have multiple decay paths. Need to review why this is the case.

In [76]:
known_isotopes = [isotope for isotope in efficiency_df.index if
    isotope not in isotopes_with_no_half_lives]
half_life_df = pd.DataFrame(data = half_lives_dict,
                            index = ["Half Life in Seconds"])
half_life_df = half_life_df.transpose()
half_life_df["Parent Isotope"] = [isotope for isotope in half_life_df.index]
half_life_df["Change in Energy Release after Neutron Absorption (keV)"] = [
    post_absorption_energy[isotope] for isotope in half_life_df.index]
half_life_df

Half Life in Seconds Parent Isotope  \
1Nn            6.139000e+02            1Nn   
4N             1.754760e-22             4N   
3H             3.887813e+08             3H   
5H             8.608259e-23             5H   
6H             2.943469e-22             6H   
...                     ...            ...   
292Lv          1.280000e-02          292Lv   
293Lv          5.300000e-02          293Lv   
293Ts          1.400000e-02          293Ts   
294Ts          5.100000e-02          294Ts   
294Og          5.800000e-04          294Og   

       Change in Energy Release after Neutron Absorption (keV)  
1Nn                                                 0.00        
4N                                                  0.00        
3H                                               2880.39        
5H                                                  0.00        
6H                                                  0.00        
...                                                  ...        
292Lv                                               0.00        
293Lv                                               0.00        
293Ts                                               0.00        
294Ts                                               0.00        
294Og                                               0.00        

[2952 rows x 3 columns]

In [79]:
fig = px.scatter(half_life_df,
  x="Half Life in Seconds",
  y = "Change in Energy Release after Neutron Absorption (keV)",
  hover_data = ["Parent Isotope"],
  title = "Efficiency: Additional Energy Released by Decay Chain after Absorbing 1 Neutron",
                 log_x = True)
fig

Now you can download the file to use locally

In [80]:
plot_file_name = "Half_life_vs_Energy_From_Neutron_Irradiation.html"
fig.write_html(plot_file_name)

Now I'll export this plot to a different Github repository so that I can use this interactively plot on the internet.

In [ ]:
'''
plots_dir = "C:\\Users\\engin\\Documents\\Github\\plots\\"
fig.write_html(plots_dir + plot_file_name)
!git -C C:/Users/engin/Documents/Github/plots add *
!git -C C:/Users/engin/Documents/Github/plots commit -m "programmatically adding plot"
!git -C C:/Users/engin/Documents/Github/plots push
'''

[main 4e254ee] programmatically adding plot
 1 file changed, 71 insertions(+)
 create mode 100644 Energy_From_Neutron_Irradiation.html


To https://github.com/MarcosP7635/plots.git
   9ebc2a0..4e254ee  main -> main


In [ ]:
livechart['Isotope']

0          1Nn
1           4N
2           6N
3           1H
4           2H
         ...  
20197    293Lv
20198    294Lv
20199    293Ts
20200    294Ts
20201    294Og
Name: Isotope, Length: 20202, dtype: object

In [ ]:
isotopes = [x for x in livechart["Isotope"]]

In [ ]:
isotope = "1Nn"
[row for row in range(len(isotopes)) if isotopes[row] == isotope]

[0, 3367, 6734, 10101, 13468, 16835]

In [ ]:
livechart['half_life_sec']

0                         613.9
1        1.7547604425822527e-22
2                              
3                              
4                              
                  ...          
20197                     0.053
20198                          
20199                     0.014
20200                     0.051
20201                   0.00058
Name: half_life_sec, Length: 20202, dtype: object

It will be accessible at the following URLs:

https://marcosp7635.github.io/plots/Energy_From_Neutron_Irradiation.html

https://mp7635plots.web.app/Energy_From_Neutron_Irradiation.html


In [ ]:
h_2_mass = 2.01410
h_3_mass = 3.01604
n_mass = 1.00866
delta_m = h_3_mass - (h_2_mass + n_mass)
delta_m

-0.006720000000000059

Thus we have a mass change in  ~ $10^{-30}$ kg or maximum energy output of ~$10^{-13}$ J or 1 MeV

In [ ]:
isotopes = [x for x in df.columns]
dict = {}
for isotope in isotopes:
   dict[isotope] = df.loc[0, isotope]
dict

{'100Ag': 1055.3,
 '100Cd': 1518.64,
 '100In': 2857.78,
 '100Kr': 17239.11,
 '100Nb': 2414,
 '100Rb': 13269.11,
 '100Rh': 40.92,
 '100Sn': 2857.78,
 '100Sr': 9001.1,
 '100Tc': 1317.28575,
 '100Y': 6470.1,
 '100Zr': 3528.1,
 '101Ag': 844.51,
 '101Cd': 1790.89,
 '101In': 1790.89,
 '101Mo': 995.88,
 '101Nb': 2792.88,
 '101Pd': 17.33,
 '101Rb': 15450.97,
 '101Sn': 1790.89,
 '101Sr': 11419.6,
 '101Tc': 473.57,
 '101Y': 7622.93,
 '101Zr': 4988.03,
 '102Ag': 0,
 '102Cd': 59.84,
 '102In': 59.84,
 '102Mo': 2284.95,
 '102Nb': 4232.95,
 '102Rb': 16207.09,
 '102Rh': 78.8048,
 '102Sn': 1207.8,
 '102Sr': 11589.09,
 '102Tc': 1935,
 '102Y': 8572.09,
 '102Zr': 5476.09,
 '103Ag': 172.35,
 '103Cd': 508.5,
 '103In': 2069.7,
 '103Mo': 2290.37,
 '103Nb': 4401.37,
 '103Rb': 17846.26,
 '103Ru': 63.64,
 '103Sn': 2333.99,
 '103Sr': 13484.26,
 '103Tc': 1040.37,
 '103Y': 9790.26,
 '103Zr': 6756.26,
 '104Ag': 76.9,
 '104Cd': 76.9,
 '104In': 573.74,
 '104Mo': 1783.24,
 '104Nb': 4068.4,
 '104Rh': 980.4679501,
 '104S